### DISCUSSION TITLE
Custom theme / plugins running on Kubernetes

### DISCUSSION THREAD

**Discussion by oxodesign** [docker-library/wordpress#623] [General]
Hi everyone, 

I got things running correctly and as I want using Docker Compose but when it comes to our test server (Kubernetes) im having issues with permission on `/var/www/html` folder.

I tried copying things to `/usr/src/wordpress` but it looks like this only works on first install! Changes that has been done on the theme are not being updated.

So can someone explain (as for **dummies**, **step by step**) how to get this up and running, my docker using Dockerfile?

My Dockerfile looks like this (very minimal):
```
FROM wordpress:latest

COPY php/ext/upload.ini /usr/local/etc/php/conf.d/upload.ini
COPY wp-content/ /var/www/html/wp-content
```

All help is appreciated 

Thanks in advance

**Comment by wglambert** [comment]
The entrypoint just copies `/usr/src/wordpress` to `/var/www/html` and essentiallty `chown`'s its contents
https://github.com/docker-library/wordpress/blob/1d90641dc2075168fe59df2f02502df068cc5531/latest/php8.0/apache/docker-entrypoint.sh#L38-L42

For the Debian variant you'll want to grant permissions on the volume you're mounting for UID 33 (www-data in the container), if you're using Alpine then UID 82
Also you can run the image as an arbitrary user https://github.com/docker-library/docs/tree/master/wordpress#running-as-an-arbitrary-user

You could also try asking over at the [Docker Community Forums](https://forums.docker.com/), [Docker Community Slack](https://dockr.ly/slack), or [Stack Overflow](https://stackoverflow.com/search?tab=newest&q=docker). Since these repos aren't really a user-help forum

**Comment by oxodesign** [comment]
> The entrypoint just copies `/usr/src/wordpress` to `/var/www/html` and essentiallty `chown`'s its contents
> https://github.com/docker-library/wordpress/blob/1d90641dc2075168fe59df2f02502df068cc5531/latest/php8.0/apache/docker-entrypoint.sh#L38-L42
> 
> For the Debian variant you'll want to grant permissions on the volume you're mounting for UID 33 (www-data in the container), if you're using Alpine then UID 82
> Also you can run the image as an arbitrary user https://github.com/docker-library/docs/tree/master/wordpress#running-as-an-arbitrary-user
> 
> You could also try asking over at the [Docker Community Forums](https://forums.docker.com/), [Docker Community Slack](https://dockr.ly/slack), or [Stack Overflow](https://stackoverflow.com/search?tab=newest&q=docker). Since these repos aren't really a user-help forum

Thanks a lot for you reply, still dont have a clue how to do this! And probably this is very easy but I still dont understand to do this. Will appriciate if you provided with a simple Dockerfile on how to do this but understand if you dont want/have time.

Btw I asked Docker Slack and Kubernetes Slack without any help :(

**Comment by EvilBox** [comment]
@oxodesign 
First, you need to find out from which user uid apache or php-fpm is running in k8s (33/82/etc?)
You can add InitContainer to your deployment.yaml. Which will move the folder to the right place for WP and assign the owner/rights (chown/chmod)
See - https://kubernetes.io/docs/concepts/workloads/pods/init-containers/

**Comment by rickyseezy** [comment]
I'm having the same issues ! 
I tried to do this : COPY ./app/wp-content/themes /usr/src/wordpress/wp-content/themes
then this : COPY ./app/wp-content/themes/my-theme /usr/src/wordpress/wp-content/themes

But still not working on Kubernetes

**Comment by yosifkit** [comment]
The entrypoint will only copy `/usr/src/wordpress/` into the current directory (default `/var/www/html`, which is also defined as a `VOLUME`) if the directory is deemed "empty" (see the entrypoint for complete logic):

https://github.com/docker-library/wordpress/blob/55bb0d5776bc236a29bc72acae4f184625d53a98/latest/php8.0/apache/docker-entrypoint.sh#L28

So if you have configured kubernetes (or even just a basic `docker run`) to use a persistent volume for `/var/www/html/`, then updating the image with a new build have no change to the wordpress install contents at `/var/www/html/`. The new image could have a newer PHP or OS packages and those would be work, but the design of the image is such that WordPress itself should be managing updates to itself and WP plugins/themes since that is how WordPress expects to be run.

If you are setting a specific uid/gid for the container to run as (like kubernetes' `securityContext.runAsUser`), then you need to also make sure that the volume is accessible for that user/group.

---

NOTE: `docker build` cannot build a useful WordPress image with a Dockerfile that has `COPY some-files /var/www/html/` since that directory is a VOLUME (the files will just be discarded and not committed to the image).

I *think* you want something like this but again, this setup assumes that you will manage WordPress and theme/plugin updates via the WP admin UI.

```dockerfile
# be specific about WP version, php version, and apache (or fpm or fpm-alpine)
FROM wordpress:5.7-php8.0-apache

# should double check that these commands put your files/folders where you expected
# and that they have the correct permissions
# ie, "docker run -it --rm my-custom-wp bash" and have a look around
COPY upload.ini /usr/local/etc/php/conf.d/upload.ini
COPY custom-theme /usr/src/wordpress/wp-content/themes/
```
